### PR TITLE
[SPARK-19014][SQL] support complex aggregate buffer in HashAggregateExec

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Currently we only support aggregate buffer with simple/primitive types for `HashAggregateExec`. Ideally the buffer type should be mutable in an unsafe row, however, logicall struct type with all fixed-length fields is also mutable.

This PR fixes this problem, and we can run whole-stage codegen for aggregate function with complex buffer type.

Did a micro benchmark:
```
val df = spark.range(100000000).select(max(struct($"id")))
df.explain
df.collect
```

Before this PR, the explain result is:
```
== Physical Plan ==
SortAggregate(key=[], functions=[max(named_struct(id, id#0L))])
+- Exchange SinglePartition
   +- SortAggregate(key=[], functions=[partial_max(named_struct(id, id#0L))])
      +- *Range (0, 100000000, step=1, splits=Some(1))
```
and it costs 4800 ms

After this PR, the explain result is:
```
== Physical Plan ==
*HashAggregate(keys=[], functions=[max(named_struct(id, id#11L))], output=[max(named_struct(NamePlaceholder(), id))#15])
+- Exchange SinglePartition
   +- *HashAggregate(keys=[], functions=[partial_max(named_struct(id, id#11L))], output=[max#28])
      +- *Range (0, 10000000, step=1, splits=Some(1))
```
and it costs 1500 ms

## How was this patch tested?

N/A

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai @hvanhovell

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -553,11 +580,36 @@ class CodegenContext {
           }
         """
       addNewFunction(compareFunc, funcCode)
-      s"this.$compareFunc($c1, $c2)"
+      s"$compareFunc($c1, $c2)"
     case schema: StructType =>
-      INPUT_ROW = "i"
-      val comparisons = GenerateOrdering.genComparisons(this, schema)
       val compareFunc = freshName("compareStruct")
+      val compareFields = schema.zipWithIndex.map {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: the previous comparison code does not work in whole-stage codegen, as it mutate the `INPUT_ROW`.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 2]
```diff
@@ -201,6 +210,25 @@ public void setNullAt(int i) {
     Platform.putLong(baseObject, getFieldOffset(i), 0);
   }
 
+  public void setNullData(int ordinal) {
```
Comment (viirya) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: Might be confused with `setNullAt`, add a short comment for this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala [hunk 3]
```diff
@@ -92,46 +89,53 @@ object GenerateUnsafeProjection extends CodeGenerator[Seq[Expression], UnsafePro
...
         }
         val tmpCursor = ctx.freshName("tmpCursor")
 
         val setNull = dt match {
-          case t: DecimalType if t.precision > Decimal.MAX_LONG_DIGITS =>
-            // Can't call setNullAt() for DecimalType with precision larger than 18.
-            s"$rowWriter.write($index, (Decimal) null, ${t.precision}, ${t.scale});"
+          case dt if UnsafeRow.isMutable(dt) && !UnsafeRow.isFixedLength(dt) =>
+            def varLenDataSize(s: DataType): Int = s match {
+              case t: DecimalType if t.precision > Decimal.MAX_LONG_DIGITS => 16
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala]: If `!UnsafeRow.isFixedLength(dt)` and `dt` is `DecimalType`, I think it already implicitly means `t.precision > Decimal.MAX_LONG_DIGITS`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 4]
```diff
@@ -310,6 +310,31 @@ class CodegenContext {
...
+        }
+        val updateFieldsCode = splitExpressions(
+          updateFields, "updateFields", Seq("InternalRow" -> nestedRow, "InternalRow" -> "value"))
+        val setColumnFunc = freshName("setColumnFunc")
+        val funcCode = s"""
+          public void $setColumnFunc(InternalRow row, InternalRow value) {
+            if (row instanceof UnsafeRow) {
+              ((UnsafeRow) row).setNotNullAt($ordinal);
+              final InternalRow $nestedRow = row.getStruct($ordinal, ${s.length});
+              $updateFieldsCode
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: A normal use case in aggregation would be that the buffer is written back. In that case the `value` row and the `nestedRow` should be the same, and then we don't have write anything (skipping the updates).

A related thought is that we could just copy the bytes from `value` to `nestedRow`.

Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Is `value` always `UnsafeRow`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala (hunk 3 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala]: yup

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala (hunk 4 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: > In that case the value row and the nestedRow should be the same

yea we should skip updating for this case.

> we could just copy the bytes from value to nestedRow.

well, we can't guarantee the `value` is `UnsafeRow` and fixed-length

**Comment by viirya** [comment]
retest this please.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java (hunk 2 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: any ideas on the method name?

Comment (viirya) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: setNullForNonPrimitive? setNullAtNonPrimitive?

Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: how about `setNullForFixedLenthNonPrimitive`?

Comment (viirya) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: Ok. Good for me.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
Jenkins looks unstable.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java [hunk 5]
```diff
@@ -303,6 +331,15 @@ public void setDecimal(int ordinal, Decimal value, int precision) {
     }
   }
 
+  public void setFixedLengthStruct(int ordinal, UnsafeRow row) {
```
Comment (viirya) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java]: Do we need call `setNotNullAt(ordinal)` here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 6]
```diff
@@ -327,24 +359,28 @@ class CodegenContext {
       nullable: Boolean,
       isVectorized: Boolean = false): String = {
     if (nullable) {
-      // Can't call setNullAt on DecimalType, because we need to keep the offset
-      if (!isVectorized && dataType.isInstanceOf[DecimalType]) {
+      // Can't call setNullAt for mutable but non-primitive data type, e.g. DecimalType, StructType
+      // with fix-length fields, because we need to keep the offset and length.
+      val isMutableVarLenField = UnsafeRow.isMutable(dataType) && !UnsafeRow.isFixedLength(dataType)
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala]: Do you think it is good to add `isMutableVarLenField` method into `UnsafeRow` directly?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala [hunk 7]
```diff
@@ -92,46 +89,53 @@ object GenerateUnsafeProjection extends CodeGenerator[Seq[Expression], UnsafePro
...
           case other => other
         }
         val tmpCursor = ctx.freshName("tmpCursor")
 
         val setNull = dt match {
-          case t: DecimalType if t.precision > Decimal.MAX_LONG_DIGITS =>
-            // Can't call setNullAt() for DecimalType with precision larger than 18.
-            s"$rowWriter.write($index, (Decimal) null, ${t.precision}, ${t.scale});"
+          case dt if UnsafeRow.isMutable(dt) && !UnsafeRow.isFixedLength(dt) =>
+            def varLenDataSize(s: DataType): Int = s match {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala]: Is it good to put this `varLenDataSize` into `UnsafeRow`? Writing a constant like 16 here makes the code untraceable for new comers.

**Comment by cloud-fan** [comment]
retest this please

**Pull request closed without merging**